src/jdk.jdwp.agent/share/native/libjdwp/log_messages.c

```diff
@@ -78,9 +78,12 @@ get_time_stamp(char *tbuf, size_t ltbuf)
                 "%d.%m.%Y %T", localtime(&t));
     (void)strftime(timestamp_timezone, TZ_SIZE,
                 "%Z", localtime(&t));
-    (void)snprintf(tbuf, ltbuf,
-                   "%s.%.3d %s", timestamp_date_time,
-                   (int)(millisecs), timestamp_timezone);
+
+    // Truncate milliseconds in buffer large enough to hold the
+    // value which is always < 1000 (and so a maximum of 3 digits for "%.3d")
+    char tmp[20];
+    snprintf(tmp, sizeof(tmp), "%.3d", millisecs);
+    snprintf(tbuf, ltbuf, "%s.%s %s", timestamp_date_time, tmp, timestamp_timezone);
```

Comment:
Sorry, Coleen, we still miss the precision limit in the second printf.

```
char tmp[20];
snprintf(tmp, sizeof(tmp), "%.3d", millisecs);
snprintf(tbuf, ltbuf, "%s.%.3s %s", timestamp_date_time, tmp, timestamp_timezone);
                          ^^^^
```

The point was to first print the digit with enough space to hold its stringified value (20 is more than enough). We need to print that one with .3d to get zero-padding to three digits if the value is <100.

Then to print that stringified version with ".3s" since, in contrast to %d, with %s the precision actually works as a limiter.

Thanks for your patience.

Reply:
Thanks for all your help and quick responses.  I made that change, see new commit.  I should realize that there are no simple changes!